src/neptune_scale/api/attribute.py

```diff
@@ -0,0 +1,243 @@
+            self._attributes[path] = attr
+
+        return attr
+
+    def __setitem__(self, key: str, value: ValueType) -> None:
+        # TODO: validate type if attr is already known
+        attr = self[key]
+        attr.assign(value)
+
+    def log(
+        self,
+        step: Optional[Union[float, int]] = None,
+        timestamp: Optional[Union[datetime, float]] = None,
+        configs: Optional[Dict[str, Union[float, bool, int, str, datetime, list, set, tuple]]] = None,
+        metrics: Optional[Dict[str, Union[float, int]]] = None,
+        tags_add: Optional[Dict[str, Union[List[str], Set[str], Tuple[str]]]] = None,
+        tags_remove: Optional[Dict[str, Union[List[str], Set[str], Tuple[str]]]] = None,
+    ) -> None:
+        # TODO: This should not call Run.log, but do the actual work. Reverse the current dependency so that this
+        #       class handles all the logging
+        timestamp = datetime.now() if timestamp is None else timestamp
+
+        # TODO: Remove this and teach MetadataSplitter to handle Nones
+        configs = {} if configs is None else configs
+        metrics = {} if metrics is None else metrics
+        tags_add = {} if tags_add is None else tags_add
+        tags_remove = {} if tags_remove is None else tags_remove
+
+        print(f"log({step}, {configs}, {metrics}, {tags_add}, {tags_remove}")
+
+        # TODO: remove once Run.log accepts Union[datetime, float]
+        timestamp = cast(datetime, timestamp)
+        self._run.log(
+            step=step, timestamp=timestamp, configs=configs, metrics=metrics, tags_add=tags_add, tags_remove=tags_remove
+        )
+
+
+class Attribute:
+    """Objects of this class are returned on dict-like access to Run. Attributes have a path and
+    allow logging values under it.
+
+    run = Run(...)
+    run['foo'] = 1
+    run['nested'] = {'foo': 1, {'bar': {'baz': 2}}}
+    run['bar'].append(1, step=10)
+    """
+
+    def __init__(self, store: AttributeStore, path: str) -> None:
+        self._store = store
+        self._path = path
+
+    # TODO: typehint value properly
+    @warn_unsupported_params
+    def assign(self, value: Any, *, wait: bool = False) -> None:
+        data = accumulate_dict_values(value, self._path)
+        self._store.log(configs=data)
+
+    @warn_unsupported_params
+    def append(
+        self,
+        value: Union[Dict[str, Any], float],
+        *,
+        step: Union[float, int],
+        timestamp: Optional[Union[float, datetime]] = None,
+        wait: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        data = accumulate_dict_values(value, self._path)
+        self._store.log(metrics=data, step=step, timestamp=timestamp)
+
+    @warn_unsupported_params
+    # TODO: this should be Iterable in Run as well
+    # def add(self, values: Union[str, Iterable[str]], *, wait: bool = False) -> None:
+    def add(self, values: Union[str, Union[List[str], Set[str], Tuple[str]]], *, wait: bool = False) -> None:
```

Comment:
is it really clear that this method adds tags if you shorten add_tags to just add()?

can't you type it as `values: Union[str, Iterable[str]]`right now and copy it to tuple `self._store.log(tags_add={self._path: tuple(values)})` at the point where you're calling run.log?

Reply:
This method name is taken from the old client, for which the API needs to be compatible. I'll add `add_/remove_tags` in another PR, deprecating this one